tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -922,6 +935,289 @@ HRESULT DoSimpleReWrite(_In_ DxcLangExtensionsHelper *pHelper,
+  compiler.getLangOpts().UseMinPrecision = !opts.Enable16BitTypes;
+  compiler.getLangOpts().EnableDX9CompatMode = opts.EnableDX9CompatMode;
+  compiler.getLangOpts().EnableFXCCompatMode = opts.EnableFXCCompatMode;
+  compiler.getDiagnostics().setIgnoreAllWarnings(!opts.OutputWarnings);
+  compiler.getCodeGenOpts().MainFileName = pMainFile;
+
+  PreprocessorOptions &PPOpts = compiler.getPreprocessorOpts();
+  if (rewrite != nullptr) {
+    if (llvm::MemoryBuffer *pMemBuf = rewrite->second) {
+      compiler.getPreprocessorOpts().addRemappedFile(StringRef(pMainFile),
+                                                     pMemBuf);
+    }
+
+    PPOpts.RemappedFilesKeepOriginalName = true;
+  }
+
+  PPOpts.ExpandTokPastingArg = opts.LegacyMacroExpansion;
+
+  // Pick additional arguments.
+  clang::HeaderSearchOptions &HSOpts = compiler.getHeaderSearchOpts();
+  HSOpts.UseBuiltinIncludes = 0;
+  // Consider: should we force-include '.' if the source file is relative?
+  for (const llvm::opt::Arg *A : opts.Args.filtered(options::OPT_I)) {
+    const bool IsFrameworkFalse = false;
+    const bool IgnoreSysRoot = true;
+    if (dxcutil::IsAbsoluteOrCurDirRelative(A->getValue())) {
+      HSOpts.AddPath(A->getValue(), frontend::Angled, IsFrameworkFalse,
+                     IgnoreSysRoot);
+    } else {
+      std::string s("./");
+      s += A->getValue();
+      HSOpts.AddPath(s, frontend::Angled, IsFrameworkFalse, IgnoreSysRoot);
+    }
+  }
+
+  if (msfPtr) {
+    msfPtr->SetupForCompilerInstance(compiler);
+  }
+
+  if (pDefines) {
+    std::string newDefines = compiler.getPreprocessor().getPredefines();
+    newDefines += pDefines;
+    compiler.getPreprocessor().setPredefines(newDefines);
+  }
+}
+
+void PreprocessResult(CompilerInstance &compiler, LPCSTR pFileName) {
+  // These settings are back-compatible with fxc.
+  clang::PreprocessorOutputOptions &PPOutOpts =
+      compiler.getPreprocessorOutputOpts();
+  PPOutOpts.ShowCPP = 1;           // Print normal preprocessed output.
+  PPOutOpts.ShowComments = 0;      // Show comments.
+  PPOutOpts.ShowLineMarkers = 1;   // Show \#line markers.
+  PPOutOpts.UseLineDirectives = 1; // Use \#line instead of GCC-style \# N.
+  PPOutOpts.ShowMacroComments = 0; // Show comments, even in macros.
+  PPOutOpts.ShowMacros = 0;        // Print macro definitions.
+  PPOutOpts.RewriteIncludes = 0;   // Preprocess include directives only.
+
+  FrontendInputFile file(pFileName, IK_HLSL);
+  clang::PrintPreprocessedAction action;
+  if (action.BeginSourceFile(compiler, file)) {
+    action.Execute();
+    action.EndSourceFile();
+  }
+}
+
+class RewriteVisitor : public RecursiveASTVisitor<RewriteVisitor> {
+public:
+  RewriteVisitor(Rewriter &R, TranslationUnitDecl *tu, RewriteHelper &helper)
+      : TheRewriter(R), SourceMgr(R.getSourceMgr()), tu(tu),
+        helper(helper), bNeedLineInfo(false) {}
+
+  bool VisitFunctionDecl(FunctionDecl *f) {
+    if (helper.unusedFunctions.count(f)) {
```

Comment:
This and those like it below certainly catches when something has been removed. I think the result of this is to inline all include files as well. Will it print a line when those are included?

Additionally, the way declprinter works, sometimes what took up a single line originally might end up on multiple lines such as: `int a = 0; int b = 2;` Does it handle that case?